op-node/rollup/derive/plasma_data_source.go

```diff
@@ -0,0 +1,52 @@
+package derive
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/ethereum-optimism/optimism/op-service/eth"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+// PlasmaDataSource is a data source that fetches inputs from a plasma DA provider given
+// their onchain commitments. Same as CalldataSource it will keep attempting to fetch.
+type PlasmaDataSource struct {
+	log     log.Logger
+	src     DataIter
+	fetcher PlasmaInputFetcher
+	id      eth.BlockID
+	// keep track of a pending commitment so we can keep trying to fetch the input.
+	comm []byte
+}
+
+func NewPlasmaDataSource(log log.Logger, src DataIter, fetcher PlasmaInputFetcher, id eth.BlockID) *PlasmaDataSource {
+	return &PlasmaDataSource{
+		log:     log,
+		src:     src,
+		fetcher: fetcher,
+		id:      id,
+	}
+}
+
+func (s *PlasmaDataSource) Next(ctx context.Context) (eth.Data, error) {
+	if s.comm == nil {
+		var err error
+		// the l1 source returns the input commitment for the batch.
+		s.comm, err = s.src.Next(ctx)
+		if err != nil {
+			return nil, err
+		}
+	}
+	// use the commitment to fetch the input from the plasma DA provider.
+	resp, err := s.fetcher.GetInput(ctx, s.comm, s.id.Number)
+	if err != nil {
+		// wrap in eth.Data for printing as hex string.
+		comm := eth.Data(s.comm)
+		// return temporary error so we can keep retrying.
+		tempErr := NewTemporaryError(fmt.Errorf("failed to fetch input data with comm %v from da service: %w", comm, err))
```

Comment:
no need for `tempErr`, can return it directly. Also, in some cases we may not want just a temporary error